--- a/drivers/net/wireless/ath/ath11k/debugfs_sta.c
+++ b/drivers/net/wireless/ath/ath11k/debugfs_sta.c
@@ -9,6 +9,27 @@
 #include "peer.h"
 #include "debug.h"
 
+static u8 mac80211_bw_to_ath11k_bw( u8 bw)
+{
+	u8 ret = 0;
+	switch (bw) {
+	case RATE_INFO_BW_20:
+		ret = ATH11K_BW_20;
+		break;
+	case RATE_INFO_BW_40:
+		ret = ATH11K_BW_40;
+		break;
+	case RATE_INFO_BW_80:
+		ret = ATH11K_BW_80;
+		break;
+	case RATE_INFO_BW_160:
+		ret = ATH11K_BW_160;
+		break;
+	}
+
+	return ret;
+}
+
 void
 ath11k_accumulate_per_peer_tx_stats(struct ath11k_sta *arsta,
 				    struct ath11k_per_peer_tx_stats *peer_stats,
@@ -24,7 +45,7 @@ ath11k_accumulate_per_peer_tx_stats(stru
 	tx_stats = arsta->tx_stats;
 	gi = FIELD_GET(RATE_INFO_FLAGS_SHORT_GI, arsta->txrate.flags);
 	mcs = txrate->mcs;
-	bw = txrate->bw;
+	bw = mac80211_bw_to_ath11k_bw(txrate->bw);
 	nss = txrate->nss - 1;
 
 #define STATS_OP_FMT(name) tx_stats->stats[ATH11K_STATS_TYPE_##name]
